File: tests/performance/cg.py
import numpy as np
import time
import sys

dim = 16384
if len(sys.argv) > 1:
    dim = int(sys.argv[1])
print(f'Matrix Dimensions: {dim}x{dim}')
print(f'Vector Dimensions: {dim}')

A = np.full((dim, dim), 1.)
b = np.full((dim), 1.)
x = np.full((dim), 1.)

for row in range(dim):
    for col in range(dim):
        A[row][col] = (row + col) % 10

for dimX in range(dim):
    b[dimX] = dimX
    x[dimX] = dimX

start = time.time()

r = b - np.dot(A, x)
p = r
rsold = np.dot(r, r)
print(f'[-1] Rsold: {rsold}')

for i in range(100):
    Ap = np.dot(A, p)
    alpha = rsold / np.dot(p, Ap)

    x = x + (alpha * p)
    r = r - (alpha * Ap)

    rsnew = np.dot(r, r)
    if (np.sqrt(rsnew) < 1E-8):
        print(f'Converged in {i} iterations!')
        break

    p = r + (rsnew / rsold) * p
    rsold = rsnew

    print(f'[{i}] Rsold: {rsold}')

end = time.time()
print(f'Execution Time: {end - start}')
print(f'Rsold: {rsold}')

File: tests/performance/cg.cpp
#include <charmtyles/charmtyles.hpp>

#include "cg.decl.h"

class cg_generator : public ct::generator
{
public:
    cg_generator() = default;
    ~cg_generator() {}

    using ct::generator::generator;

    double generate(int row_id, int col_id) final
    {
        return (row_id + col_id) % 10;
    }

    double generate(int dimX) final
    {
        return dimX;
    }

    PUPable_decl(cg_generator);
    cg_generator(CkMigrateMessage* m)
      : ct::generator(m)
    {
    }

    void pup(PUP::er& p) final
    {
        ct::generator::pup(p);
    }
};

class Main : public CBase_Main
{
public:
    Main(CkArgMsg* msg)
    {
        int dim = 1 << 14;
        if (msg->argc > 1)
            dim = atoi(msg->argv[1]);

        ct::init();
        thisProxy.benchmark(dim);
    }

    void benchmark(int dim)
    {
        ckout << "Matrix Dimensions: " << dim << "x" << dim << endl;
        ckout << "Vector Dimensions: " << dim << endl;

        ct::vec_impl::vec_instr_queue_t& vec_queue =
            CT_ACCESS_SINGLETON(ct::vec_impl::vec_instr_queue);
        ct::mat_impl::mat_instr_queue_t& mat_queue =
            CT_ACCESS_SINGLETON(ct::mat_impl::mat_instr_queue);

        std::shared_ptr<cg_generator> new_cg_generator =
            std::make_shared<cg_generator>();

        ct::matrix A{dim, dim, new_cg_generator};

        // Random b, x
        ct::vector b{dim, new_cg_generator};
        ct::vector x{dim, new_cg_generator};

        ct::sync();

        double start = CkWallTimer();

        ct::vector a_x = ct::dot(A, x);
        ct::scalar dax = ct::dot(a_x, a_x);
        ct::vector r = b - a_x;
        ct::vector p = r;

        ct::scalar rsold = ct::dot(r, r);
        ckout << "[-1] Rsold: " << rsold.get() << endl;

        double gres = 0.;
        ct::vector Ap = ct::dot(A, p);
        ct::scalar pAp = ct::dot(p, Ap);
        double rsold_value = rsold.get();
        double alpha = rsold_value / pAp.get();

        x = ct::axpy(alpha, p, x);
        r = ct::axpy(-alpha, Ap, r);

        ct::scalar rsnew = ct::dot(r, r);
        double rsnew_value = rsnew.get();

        if (std::sqrt(rsnew_value) < 1E-8)
            ckout << "Converged in 0 iterations" << endl;

        p = ct::axpy(rsnew_value / rsold_value, p, r);
        rsold = rsnew;

        ckout << "[0] Rsold: " << rsold.get() << endl;

        for (int i = 1; i < 100; ++i)
        {
            Ap = ct::dot(A, p);
            pAp = ct::dot(p, Ap);
            rsold_value = rsold.get();
            alpha = rsold_value / pAp.get();

            x = ct::axpy(alpha, p, x);
            r = ct::axpy(-alpha, Ap, r);

            rsnew = ct::dot(r, r);
            rsnew_value = rsnew.get();

            if (std::sqrt(rsnew_value) < 1E-8)
            {
                ckout << "Converged in " << i << " iterations" << endl;
                break;
            }

            p = ct::axpy(rsnew_value / rsold_value, p, r);
            rsold = rsnew;

            ckout << "[" << i << "] Rsold: " << rsold.get() << endl;
        }

        ct::sync();
        double end = CkWallTimer();

        ckout << "Execution Time: " << end - start << endl;
        ckout << "Value: " << rsold.get() << endl;

        CkExit();
    }
};

#include "cg.def.h"
